File: integrations/dotnet-ariada/src/Ariada.DotNet.Core/AriadaCliRunner.cs
// SPDX-FileCopyrightText: 2026 Agonist Development AB
// SPDX-License-Identifier: EUPL-1.2

using System.Diagnostics;
using System.Net;
using System.Net.Sockets;

namespace Ariada.DotNet.Core;

public interface IAriadaProcessRunner
{
 Task<ProcessResult> RunAsync(IReadOnlyList<string> command, CancellationToken cancellationToken);
}

public sealed record ProcessResult(int ExitCode, string StandardOutput, string StandardError);

public sealed class SystemProcessRunner: IAriadaProcessRunner
{
 public async Task<ProcessResult> RunAsync(IReadOnlyList<string> command, CancellationToken cancellationToken)
 {
 if (command.Count == 0)
 {
 throw new ArgumentException("Command must not be empty.", nameof(command));
 }

 var start = new ProcessStartInfo
 {
 FileName = command[0],
 RedirectStandardOutput = true,
 RedirectStandardError = true,
 UseShellExecute = false,
 };

 foreach (var arg in command.Skip(1))
 {
 start.ArgumentList.Add(arg);
 }

 using var process = Process.Start(start) ?? throw new InvalidOperationException("Could not start Ariada CLI.");
 var stdout = process.StandardOutput.ReadToEndAsync(cancellationToken);
 var stderr = process.StandardError.ReadToEndAsync(cancellationToken);
 await process.WaitForExitAsync(cancellationToken).ConfigureAwait(false);
 return new ProcessResult(process.ExitCode, await stdout.ConfigureAwait(false), await stderr.ConfigureAwait(false));
 }
}

public sealed class AriadaCliRunner
{
 private readonly IAriadaProcessRunner processRunner;

 public AriadaCliRunner(IAriadaProcessRunner? processRunner = null)
 {
 this.processRunner = processRunner ?? new SystemProcessRunner();
 }

 public async Task<AriadaScanResult> RunAsync(AriadaOptions options, CancellationToken cancellationToken = default)
 {
 Directory.CreateDirectory(options.OutputDirectory);
 await using var served = LocalStaticTarget.TryServe(options.Target, cancellationToken);
 var command = BuildCommand(options with { Target = served.Url });
 var completed = await processRunner.RunAsync(command, cancellationToken).ConfigureAwait(false);
 var reportPath = FindReport(options.OutputDirectory);
 var findings = reportPath is null ? Array.Empty<AriadaFinding>(): AriadaReportParser.ParseReportFile(reportPath);

 return new AriadaScanResult(
 options.Target,
 completed.ExitCode,
 completed.StandardOutput,
 completed.StandardError,
 reportPath,
 findings);
 }

 public static IReadOnlyList<string> BuildCommand(AriadaOptions options)
 {
 var command = new List<string>
 {
 options.CliCommand,
 "scan",
 options.Target,
 "--format",
 options.Format,
 "--output-dir",
 options.OutputDirectory,
 "--browser",
 options.Browser,
 "--severity-threshold",
 options.SeverityThreshold,
 "--timeout-ms",
 options.TimeoutMilliseconds.ToString(),
 };

 if (options.Domains is { Count: > 0 })
 {
 command.Add("--domains");
 command.Add(string.Join(",", options.Domains));
 }

 return command;
 }

 private static string? FindReport(string outputDirectory)
 {
 var multi = Path.Combine(outputDirectory, "multi-domain-report.json");
 if (File.Exists(multi))
 {
 return multi;
 }

 var single = Path.Combine(outputDirectory, "scan.json");
 return File.Exists(single) ? single: null;
 }
}

internal sealed class ServedTarget: IAsyncDisposable
{
 private readonly HttpListener? listener;
 private readonly Task? serverTask;

 public ServedTarget(string url, HttpListener? listener = null, Task? serverTask = null)
 {
 Url = url;
 this.listener = listener;
 this.serverTask = serverTask;
 }

 public string Url { get; }

 public async ValueTask DisposeAsync()
 {
 if (listener is null)
 {
 return;
 }

 listener.Stop();
 listener.Close();
 if (serverTask is not null)
 {
 try
 {
 await serverTask.ConfigureAwait(false);
 }
 catch (HttpListenerException)
 {
 }
 catch (ObjectDisposedException)
 {
 }
 }
 }
}

internal static class LocalStaticTarget
{
 public static ServedTarget TryServe(string target, CancellationToken cancellationToken)
 {
 if (!Directory.Exists(target))
 {
 return new ServedTarget(target);
 }

 var root = Path.GetFullPath(target);
 var port = ReserveLoopbackPort();
 var url = $"http://127.0.0.1:{port}/";
 var listener = new HttpListener();
 listener.Prefixes.Add(url);
 listener.Start();
 var task = Task.Run(() => ServeAsync(listener, root, cancellationToken), cancellationToken);
 return new ServedTarget(url, listener, task);
 }

 private static async Task ServeAsync(HttpListener listener, string root, CancellationToken cancellationToken)
 {
 while (listener.IsListening && !cancellationToken.IsCancellationRequested)
 {
 var context = await listener.GetContextAsync().ConfigureAwait(false);
 _ = Task.Run(() => RespondAsync(context, root), cancellationToken);
 }
 }

 private static async Task RespondAsync(HttpListenerContext context, string root)
 {
 var raw = context.Request.Url?.AbsolutePath.TrimStart('/') ?? "";
 var relative = string.IsNullOrWhiteSpace(raw) ? "index.html": Uri.UnescapeDataString(raw);
 var candidate = Path.GetFullPath(Path.Combine(root, relative));
 if (!candidate.StartsWith(root, StringComparison.Ordinal) || !File.Exists(candidate))
 {
 context.Response.StatusCode = 404;
 context.Response.Close();
 return;
 }

 var bytes = await File.ReadAllBytesAsync(candidate).ConfigureAwait(false);
 context.Response.ContentType = ContentTypeFor(candidate);
 context.Response.ContentLength64 = bytes.Length;
 await context.Response.OutputStream.WriteAsync(bytes).ConfigureAwait(false);
 context.Response.Close();
 }

 private static int ReserveLoopbackPort()
 {
 var listener = new TcpListener(IPAddress.Loopback, 0);
 listener.Start();
 var port = ((IPEndPoint)listener.LocalEndpoint).Port;
 listener.Stop();
 return port;
 }

 private static string ContentTypeFor(string path)
 {
 return Path.GetExtension(path).ToLowerInvariant() switch
 {
 ".html" => "text/html; charset=utf-8",
 ".css" => "text/css; charset=utf-8",
 ".js" => "text/javascript; charset=utf-8",
 ".json" => "application/json; charset=utf-8",
 ".png" => "image/png",
 ".jpg" or ".jpeg" => "image/jpeg",
 ".svg" => "image/svg+xml",
 _ => "application/octet-stream",
 };
 }
}

File: integrations/symfony-ariada/src/Scanner/AriadaCliRunner.php
<?php

declare(strict_types=1);

namespace Ariada\Symfony\Scanner;

use Symfony\Component\Process\Process;

final class AriadaCliRunner implements AriadaScanner
{
 /**
 * @param null|callable(list<string>): ProcessResult $processRunner
 */
 public function __construct(
 private mixed $processRunner = null,
) {
 }

 public function scan(string $url, ScanOptions $options): ScanResult
 {
 if (!is_dir($options->outputDir)) {
 mkdir($options->outputDir, 0775, true);
 }

 $command = [
...$this->splitCommand($options->cliCommand),
 'scan',
 $url,
 '--format',
 $options->format,
 '--output-dir',
 $options->outputDir,
 '--browser',
 $options->browser,
 '--severity-threshold',
 $options->severityThreshold,
 '--timeout-ms',
 (string) $options->timeoutMs,
 ];

 if ($options->domains !== []) {
 $command[] = '--domains';
 $command[] = implode(',', $options->domains);
 }

 $completed = $this->runProcess($command);
 [$reportPath, $totalFindings] = $this->readReportSummary($options->outputDir);

 return new ScanResult(
 target: $url,
 exitCode: $completed->exitCode,
 stdout: $completed->stdout,
 stderr: $completed->stderr,
 reportPath: $reportPath,
 totalFindings: $totalFindings,
);
 }

 /**
 * @param list<string> $command
 */
 private function runProcess(array $command): ProcessResult
 {
 if ($this->processRunner !== null) {
 return ($this->processRunner)($command);
 }

 $process = new Process($command);
 $process->run();

 return new ProcessResult(
 exitCode: $process->getExitCode() ?? 3,
 stdout: $process->getOutput(),
 stderr: $process->getErrorOutput(),
);
 }

 /**
 * @return list<string>
 */
 private function splitCommand(string $command): array
 {
 $parts = preg_split('/\s+/', trim($command));

 return array_values(array_filter($parts ?: ['ariada'], static fn (string $part): bool => $part !== ''));
 }

 /**
 * @return array{0: null|string, 1: int}
 */
 private function readReportSummary(string $outputDir): array
 {
 foreach (['multi-domain-report.json', 'scan.json'] as $name) {
 $path = rtrim($outputDir, DIRECTORY_SEPARATOR). DIRECTORY_SEPARATOR. $name;
 if (!is_file($path)) {
 continue;
 }
 $data = json_decode((string) file_get_contents($path), true);

 return [$path, is_array($data) ? $this->countFindings($data): 0];
 }

 return [null, 0];
 }

 /**
 * @param array<string, mixed> $data
 */
 private function countFindings(array $data): int
 {
 if (isset($data['summary']['total']) && is_int($data['summary']['total'])) {
 return $data['summary']['total'];
 }

 if (!isset($data['grid']) || !is_array($data['grid'])) {
 return 0;
 }

 $total = 0;
 foreach ($data['grid'] as $site) {
 if (!is_array($site)) {
 continue;
 }
 foreach ($site as $findings) {
 if (is_array($findings)) {
 $total += count($findings);
 }
 }
 }

 return $total;
 }
}
